okio/src/test/java/okio/BufferedSourceTest.java

```diff
@@ -468,4 +469,115 @@
     } catch (ArrayIndexOutOfBoundsException expected) {
     }
   }
+
+  @Test public void longHexString() throws IOException {
+    assertLongHexString("8000000000000000");
+    assertLongHexString("fffffffffffffffe");
+    assertLongHexString("FFFFFFFFFFFFFFFe");
+    assertLongHexString("ffffffffffffffff");
+    assertLongHexString("FFFFFFFFFFFFFFFF");
+    assertLongHexString("0000000000000000");
+    assertLongHexString("0000000000000001");
+    assertLongHexString("7999999999999999");
+
+    assertLongHexString("FF");
+    assertLongHexString("01");
+  }
+
+  private void assertLongHexString(String s) throws IOException {
+    data.writeUtf8(s);
+    long expected = new BigInteger(s, 16).longValue();
+    long actual = source.readHexadecimalUnsignedLong();
+    assertEquals(s + " --> " + expected, expected, actual);
+  }
+
+  @Test public void longHexStringAcrossSegment() throws IOException {
+    data.writeUtf8(repeat('a', Segment.SIZE - 8)).writeUtf8("FFFFFFFFFFFFFFFF");
+    source.skip(Segment.SIZE - 8);
+    assertEquals(-1, source.readHexadecimalUnsignedLong());
+  }
+
+  @Test public void longHexStringTooLongThrows() throws IOException {
+    try {
+      data.writeUtf8("fffffffffffffffff");
+      source.readHexadecimalUnsignedLong();
+      fail();
+    } catch (NumberFormatException e) {
+      assertEquals("Number too large: fffffffffffffffff", e.getMessage());
+    }
+  }
+
+  @Test public void longHexStringTooShortThrows() throws IOException {
+    try {
+      data.writeUtf8(" ");
+      source.readHexadecimalUnsignedLong();
+      fail();
+    } catch (NumberFormatException e) {
+      assertEquals("Expected leading hex character but was 0x20", e.getMessage());
```

Comment:
this error message is confusing, though I don't have a better suggestion

(complaining that 0x20 isn't hex)